src/sentry/api/endpoints/organization_events_trace.py

```diff
@@ -1,36 +1,211 @@
 import logging
 from collections import OrderedDict, defaultdict, deque
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Deque,
+    Dict,
+    Iterable,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+    cast,
+)
 
 import sentry_sdk
-from django.http import Http404
+from django.http import Http404, HttpRequest, HttpResponse
 from rest_framework.response import Response
 from sentry_relay.consts import SPAN_STATUS_CODE_TO_NAME
 
 from sentry import eventstore, features
 from sentry.api.bases import NoProjects, OrganizationEventsV2EndpointBase
 from sentry.api.serializers.models.event import get_tags_with_meta
+from sentry.eventstore.models import Event
+from sentry.models import Organization
 from sentry.snuba import discover
 from sentry.utils.snuba import Dataset, SnubaQueryParams, bulk_raw_query
 from sentry.utils.validators import INVALID_EVENT_DETAILS, is_event_id
 
-logger = logging.getLogger(__name__)
-MAX_TRACE_SIZE = 100
-NODESTORE_KEYS = ["timestamp", "start_timestamp"]
+logger: logging.Logger = logging.getLogger(__name__)
+MAX_TRACE_SIZE: int = 100
+
+if TYPE_CHECKING:
+    from mypy_extensions import TypedDict
+else:
+
+    def TypedDict(*args, **kwargs):
+        pass
+
+
+_T = TypeVar("_T")
+NodeSpans = List[Dict[str, Any]]
+SnubaTransaction = TypedDict(
+    "SnubaTransaction",
+    {
+        "id": str,
+        "transaction.status": int,
+        "transaction.op": str,
+        "transaction.duration": int,
+        "transaction": str,
+        "timestamp": str,
+        "trace.span": str,
+        "trace.parent_span": str,
+        "root": str,
+        "project.id": int,
+        "project": str,
+    },
+)
+SnubaError = TypedDict(
+    "SnubaError",
+    {
+        "id": str,
+        "timestamp": str,
+        "trace.span": str,
+        "transaction": str,
+        "issue.id": int,
+        "title": str,
+        "tags[level]": str,
+        "project.id": int,
+        "project": str,
+    },
+)
+TraceError = TypedDict(
+    "TraceError",
+    {
+        "event_id": str,
+        "issue_id": int,
+        "span": str,
+        "project_id": int,
+        "project_slug": str,
+        "title": str,
+        "level": str,
+    },
+)
+LightResponse = TypedDict(
+    "LightResponse",
+    {
+        "event_id": str,
+        "span_id": str,
+        "transaction": str,
+        "transaction.duration": int,
+        "transaction.op": str,
+        "project_id": int,
+        "project_slug": str,
+        "parent_span_id": Optional[str],
+        "parent_event_id": Optional[str],
+        "generation": Optional[int],
+        "errors": List[TraceError],
+    },
+)
+FullResponse = TypedDict(
+    "FullResponse",
+    {
+        "event_id": str,
+        "span_id": str,
+        "transaction": str,
+        "transaction.duration": int,
```

Comment:
Is there a way to get keys with periods to work with the class based TypedDict? because the FullResponse is identical to the Light Response except for a few extra keys.

Reply:
What's the problem with keys with periods?

Reply:
a class-based typeddict uses python identifiers to define keys, and `.` is not valid in identifiers

Reply:
Ah, thank you. @wmak They work in 3.8, so maybe it has to stay this way for now. One thing that _might_ work, is if all the keys that were in FullResponse that are NOT part of LightResponse were valid identifiers, you could write LightResponse using the function method and then use FullResponse extending LightResponse. I'm not sure that would work though.